### PR TITLE
Estimate VAD by weighted average

### PR DESCRIPTION
**Pull request by drevicko**
Does a weighted average of centroids.

If intensity sums to zero for a category, a 'neutral' category is used or 0 if it's not present. I'm not 100% sure this is the best approach, and the name of the "neutral" category perhaps should use some convention?

Note that if there are no categories present, then no VAD (or other dimensional) estimate is returned. It may be better to use the neutral centroid if it's present in this case also.

### CODE REVIEW

**Comment by balkian** [comment]
LGTM, except for some formatting/style issues.

The neutral category can be aliased, so it would be different in every case.
However, I'm not sold on using a neutral emotion, because I haven't seen many references to it, even in dimensional models. (e.g. there's no `neutral` center in some dimensional models)

**Comment by drevicko** [comment]
Yeah, that makes sense. Essentially we leave it up to conversion creators to deal with 'neutral' as a category in a reasonable way. 

Something we could consider is for each dimensional model to have encoded a way to interpret 'neutral' emotion. This is probably an open research question in emotion research, but needs to be addressed if we want to convert in a principled way.

I've added a commit that removes explicit treatment of "neutral" (and fixes an error I made - it shouldn't have passed checks, an attempt to convert should have raised a KeyError with my previous commit! Sounds like a new check is needed. I can have a go next week or if you've already got things setup and it's trivial for you, please go ahead).

**Comment by drevicko** [comment]
On another important note: where do we get the VAD centroids from, and how can we make one for Surprise? (currently there's only the 5 used in WN-Affect)

There's that new annotated resource where SemEval07 task 14 was annotated with VAD - we could use that?

**Comment by balkian** [comment]
We get the centroids with ANEW+WNAffect following this method: http://www.aclweb.org/anthology/W10-0208

AFAIR we discarded surprise because we didn't have enough data. We could use a similar approach to get centroids with the SemEval dataset.

As for the checks, the main difference seems to be what keys get added to totalIntensities. In that case, the defaultdict is masking those `KeyErrors`. Once we establish how to properly do this centroid conversion, we could add a test with expected values.

File: senpy/plugins/conversion/centroids.py [hunk 1]
```diff
@@ -37,14 +38,22 @@ def __init__(self, info):
...
             if category in self.centroids:
+                totalIntensities[category] += intensity
                 for dim, value in self.centroids[category].items():
                     try:
-                        res[dim] += value
+                        res[dim] += value * intensity
                     except Exception:
-                        res[dim] = value
+                        res[dim] = value * intensity
+        for dim,intensity in totalIntensities.items():
```
Comment (drevicko) [inline senpy/plugins/conversion/centroids.py]: @balkian this is where the test should have been falling down: `totalIntensities` was populated with `category` keys, but here it's keys are used in `res`, which is populated with `dim` keys. `res` should have been raising `keyError`s..

It's fixed in the current commit, but this one also passed the tests

**Comment by drevicko** [comment]
One thing to note in that paper is that they "manually define neutral to be (5,5,5)", so it makes sense to have a neutral entry among the centroids. We'd need to have some way to uniquely identify it as the 'neutral' centoid, as it's treatment is different, even if it has an alias.

Here's a go at a principled way to use it:

- re-centre the centroids around 'neutral' (so that 'neutral' is at `(0,0,0,)` )
- given intensities range from `0` to `max_intensity` and assuming each centroid is in the middle of that intensity range, scale the re-centered centroids for detected emotions by `2*intensity/max_intensity`
- add up the resulting vectors and add in neutral
- the resulting calculation for each detected emotion category looks something like:\
`result = (centroid - neutral_centroid)*2*intensity/max_intensity + neutral_centroid`

I don't like this formula if we have binary intensities (present or absent): it would effectively treat any detected emotion as having intensity `2`. Not sure how to combat that without adding a "onyx_binaryIntensity" label or some such.

Here we're assuming intensity relates linearly to changes in VAD space. I guess that's as good a first approximation as we'll get.

I think that if emotion categories are missing, it's easy: either a few centroids are left out or there are no detected emotions and we add no VAD estimate.

**Comment by drevicko** [comment]
As for surprise, let's have a go at the new data. I think doing it at a phrase/text level is more valid than averaging words as Kim and Valutti did, anyway, especially if the texts can be seen as some sort of representative sample. Seems to me they should have at least done a weighted average using WordNet's word sense frequency metadata.

**Comment by drevicko** [comment]
for reference:

https://github.com/JULIELab/EmoBank

**Comment by drevicko** [comment]
A relevant email conversation:

From @drevicko: I need to know the range of possible intensity values that the plugin generates (or assume it’s from 0 to 1 if not specified ).

Response from @balkian:
Yes and no. The way we solve it right now is through onyx:maxIntensityValue and minEmotionValue and maxEmotionValue. See https://github.com/gsi-upm/senpy/blob/master/senpy/plugins/__init__.py

However, that doesn't take into consideration different dimensions. For dimensions it would be sth like:
:MyAnalysis onyx:usesModel :DimensionModel

:DimensionModel a onyx:EmotionModel;
                         onyx:hasDimension :Dimension1 .
                                 
:Dimension1 onyx:minValue -1 ;
                   onyx:maxValue 1 .

So it would be embedded in the model itself.

**Comment by drevicko** [comment]
I don't see any use yet of "maxEmotionValue", and I'm not convinced it makes any sense for categorical emotions (how do you have negative anger?), so I guess it's more intended for dimensional representations? Strikes me that the __init__() of an EmotionPlugin isn't the right place to put this though. Better to have it in the model itself as you suggest.

As for using the max/min dimension values, I've not been taking that into account yet - for the to-VAD direction, the only reasonable approach I see is to constrain the result of conversion to VAD. This would be necessary if we adopt my approach of increasing the centroid vector by a factor of two if the intensity is at it's maximum. That's rather an arbitrary choice, maybe it's better to put the centroid at max intensity (and scale it down if intensity is less).

It would be good to add a default onyx:maxIntensityValue for categorical models, probably set to 1.
Is there an onyx:minIntensityValue? Or is this presumed to be zero?

**Comment by drevicko** [comment]
At this point, I don't check resulting dimensional values, meaning that values outsied onyx:minValue and onyx:maxValue may result (even negative results).

**Comment by drevicko** [comment]
Travis is failing to pull the senpy docker image as the uri is apparently incorrect.

**Comment by drevicko** [comment]
I presume this did pass the tests in GitLab?

**Pull request merged**